consensus/istanbul/backend/announce.go

```diff
@@ -17,405 +17,671 @@
+	for remoteAddress := range sb.lastAnnounceGossiped {
+		if !regAndElectedVals[remoteAddress] {
+			logger.Trace("Deleting entry from lastAnnounceGossiped", "address", remoteAddress, "gossip timestamp", sb.lastAnnounceGossiped[remoteAddress])
+			delete(sb.lastAnnounceGossiped, remoteAddress)
+		}
+	}
+
+	if err := sb.valEnodeTable.PruneEntries(regAndElectedVals); err != nil {
+		logger.Trace("Error in pruning valEnodeTable", "err", err)
+		return err
+	}
+
+	return nil
 }
 
-// ==============================================
+// ===============================================================
 //
-// define the functions that needs to be provided for rlp Encoder/Decoder.
+// define the IstanbulAnnounce message format, the AnnounceMsgCache entries, the announce send function (both the gossip version and the "retrieve from cache" version), and the announce get function
+
+// EnodeURL ciphertext encrypted with the public key associated with the DecryptorAddress field
+type encryptedEnode struct {
+	DecrypterAddress  common.Address
+	EncryptedEnodeURL []byte
+}
 
-// EncodeRLP serializes ar into the Ethereum RLP format.
-func (ar *announceRecord) EncodeRLP(w io.Writer) error {
-	return rlp.Encode(w, []interface{}{ar.DestAddress, ar.EncryptedEnodeURL})
+// EncodeRLP serializes ee into the Ethereum RLP format.
+func (ee *encryptedEnode) EncodeRLP(w io.Writer) error {
+	return rlp.Encode(w, []interface{}{ee.DecrypterAddress, ee.EncryptedEnodeURL})
 }
 
-// DecodeRLP implements rlp.Decoder, and load the ar fields from a RLP stream.
-func (ar *announceRecord) DecodeRLP(s *rlp.Stream) error {
+// DecodeRLP implements rlp.Decoder, and load the ee fields from a RLP stream.
+func (ee *encryptedEnode) DecodeRLP(s *rlp.Stream) error {
 	var msg struct {
-		DestAddress       common.Address
+		DecrypterAddress  common.Address
 		EncryptedEnodeURL []byte
 	}
 
 	if err := s.Decode(&msg); err != nil {
 		return err
 	}
-	ar.DestAddress, ar.EncryptedEnodeURL = msg.DestAddress, msg.EncryptedEnodeURL
+	ee.DecrypterAddress, ee.EncryptedEnodeURL = msg.DecrypterAddress, msg.EncryptedEnodeURL
 	return nil
 }
 
-// EncodeRLP serializes ad into the Ethereum RLP format.
-func (ad *announceData) EncodeRLP(w io.Writer) error {
-	return rlp.Encode(w, []interface{}{ad.AnnounceRecords, ad.EnodeURLHash, ad.Timestamp})
+func (ee *encryptedEnode) String() string {
+	return fmt.Sprintf("{DecrypterAddress: %s, EncryptedEnodeURL length: %d}", ee.DecrypterAddress.String(), len(ee.EncryptedEnodeURL))
+}
+
+// The set of encrypted enodes of ValAddress's enodeURL.  There is an encrypted enode for each registered validator.
+// The RLP encoding of this struct is the payload for the IstanbulAnnounce message.
+type validatorEncryptedEnodes struct {
+	ValAddress      common.Address
+	EncryptedEnodes []*encryptedEnode
+	EnodeURLHash    common.Hash
+	Version         uint64 // This fields first version will have a value of the ValAddress's local timestamp at the time it sent it's announce message
```

Comment:
I was a bit confused about what this comment meant on first reading. It could be simplified to 'Local Unix timestamp is used for the version number' or no?